compiler/src/iree/compiler/Codegen/Common/CPU/CPUPropagateDataLayout.cpp

```diff
@@ -0,0 +1,174 @@
+
+namespace {
+
+/// Sinks down tensor.collapse_shape across linalg.unpack op, if the collapsing
+/// dims are two unit dims where one is outer dimension and the other is inner
+/// dimension. It implies that we swap two operations by adjusting the packing
+/// metadata in linalg.unpack op.
+/// Note that the pattern only supports the case where the destination tensor of
+/// linalg.unpack op is a tensor.empty op. The constraint can be removed by
+/// introducing tensor.expand_shape op on the destination tensor. However, it is
+/// not common in practice, so it is not supported now.
+struct SinkDownCollapsingUnitDimsAcrossUnpack final
+    : public OpRewritePattern<linalg::UnPackOp> {
+  using OpRewritePattern<linalg::UnPackOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(linalg::UnPackOp op,
+                                PatternRewriter &rewriter) const override {
+    if (!isIdentityPermutation(op.getOuterDimsPerm())) {
+      return rewriter.notifyMatchFailure(
+          op, "expected identity (or unset) outer permutation");
+    }
+    if (op.getSourceRank() != op.getDestRank() + 1) {
+      return rewriter.notifyMatchFailure(
+          op, "expected unpacking exactly one dimension");
+    }
+    auto emptyOp = op.getDest().getDefiningOp<tensor::EmptyOp>();
+    if (!emptyOp) {
+      return rewriter.notifyMatchFailure(
+          op, "expected destination to be a tensor.empty op");
+    }
+    auto collapseOp = op.getSource().getDefiningOp<tensor::CollapseShapeOp>();
+    if (!collapseOp) {
+      return rewriter.notifyMatchFailure(
+          op, "expected the source to be a tensor.collapse_shape op");
+    }
+
+    SmallVector<ReassociationIndices, 4> ri =
+        collapseOp.getReassociationIndices();
+    ReassociationIndices outerRi, innerRi;
+    for (ArrayRef<int64_t> indices : ri) {
+      if (indices.size() == 1) {
+        continue;
+      }
+      if (indices.size() > 2) {
+        return rewriter.notifyMatchFailure(
+            op, "expected re-association map to have two dimensions");
+      }
+      if (outerRi.empty()) {
+        outerRi.assign(indices.begin(), indices.end());
+        continue;
+      }
+      if (innerRi.empty()) {
+        innerRi.assign(indices.begin(), indices.end());
+        continue;
+      }
+      return rewriter.notifyMatchFailure(
+          op, "expected only two re-association map to have two dimensions");
+    }
+    if (outerRi.empty() || innerRi.empty()) {
+      return rewriter.notifyMatchFailure(
+          op, "expected only two re-association map to have two dimensions");
+    }
+
+    RankedTensorType srcType = collapseOp.getSrcType();
+    if (innerRi.back() != srcType.getRank() - 1) {
+      return rewriter.notifyMatchFailure(
+          op, "expected two innermost dimensions are collapsed");
+    }
+    SmallVector<int64_t> innerDimPos(op.getInnerDimsPos());
+    if (innerDimPos[0] != outerRi[0] && outerRi[1] != innerDimPos[0]) {
```

Comment:
```suggestion
    if (innerDimPos[0] != outerRi[0] && innerDimPos[0] != outerRi[1]) {
```
I think re-arranging is more readable here.

Reply:
Thanks, I missed this. I'll update it to `!llvm::is_contained(outerRi, innerDimPos[0])`.